File: C#-Solutions/ValidateBinarySearchTree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool IsValidBST(TreeNode root) {
        return isValidBSTHelper(root, null, null);
    }

    private bool isValidBSTHelper(TreeNode root, Nullable<int> largest_right, Nullable<int> smallest_left) {
        if (root == null || (root.left == null && root.right == null)) return true;
        else if (root.left != null && root.right == null) {
            if (smallest_left == null || smallest_left > root.val) {
                smallest_left = root.val;
            }

            if (!validateRoot(root.left, largest_right, smallest_left)) return false;
            return isValidBSTHelper(root.left, largest_right, smallest_left);
        }
        else if (root.right != null && root.left == null) {
            if (largest_right == null || largest_right < root.val) {
                largest_right = root.val;
            }

            if (!validateRoot(root.right, largest_right, smallest_left)) return false;
            return isValidBSTHelper(root.right, largest_right, smallest_left);
        }
        else {
            Nullable<int> temp_l = smallest_left;

            if (smallest_left == null || smallest_left > root.val) {
                smallest_left = root.val;
            }

            if (!validateRoot(root.left, largest_right, smallest_left)) return false;
            else if (!isValidBSTHelper(root.left, largest_right, smallest_left)) return false;

            if (largest_right == null || largest_right < root.val) {
                largest_right = root.val;
            }

            if (!validateRoot(root.right, largest_right, temp_l)) return false;
            return isValidBSTHelper(root.right, largest_right, temp_l);
        }
    }

    private bool validateRoot(TreeNode root, Nullable<int> largest_right, Nullable<int> smallest_left) {
        return (largest_right == null || largest_right < root.val) && (smallest_left == null || smallest_left > root.val);
    }
}

File: PythonSolutions/ValidateBinarySearchTree.py
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        return self._isValidHelper(root, None, None)

    def _isValidHelper(self, root: Optional[TreeNode], largest_right: Optional[int], smallest_left: Optional[int]):
        if root is None or (root.left is None and root.right is None):
            return True
        elif root.left is not None and root.right is None:
            if smallest_left is None or smallest_left > root.val:
                smallest_left = root.val
            if not self._validate_root(root.left, largest_right, smallest_left):
                return False
            return self._isValidHelper(root.left, largest_right, smallest_left)
        elif root.right is not None and root.left is None:
            if largest_right is None or largest_right < root.val:
                largest_right = root.val
            if not self._validate_root(root.right, largest_right, smallest_left):
                return False
            return self._isValidHelper(root.right, largest_right, smallest_left)
        else:
            temp_l = smallest_left

            if smallest_left is None or smallest_left > root.val:
                smallest_left = root.val
            if not self._validate_root(root.left, largest_right, smallest_left):
                return False
            elif not self._isValidHelper(root.left, largest_right, smallest_left):
                return False

            if largest_right is None or largest_right < root.val:
                largest_right = root.val
            if not self._validate_root(root.right, largest_right, temp_l):
                return False
            return self._isValidHelper(root.right, largest_right, temp_l)

    def _validate_root(self, root: TreeNode, largest_right: Optional[int], smallest_left: Optional[int]):
        return (largest_right is None or root.val > largest_right) and (
                    smallest_left is None or root.val < smallest_left)
